### PR TITLE
librbd: only mark async completion as done after callback

### PR DESCRIPTION
**Pull request by marcan**
This change in behavior was introduced in fde9f78.

cc @dillaman

### CODE REVIEW

**Comment by dillaman** [comment]
I remember why I had to move it -- there is a rare possible race condition with completing no-op IO.  How about we switch ```AioCompletion::done``` to be an enum like:

```
enum State {
  STATE_PENDING,
  STATE_CALLBACK,
  STATE_COMPLETE
};
```

The ```AioCompletion::wait_for_complete``` would loop until the state is COMPLETE. Other references to ```done``` would only check for ```state != STATE_PENDING```.  The ```AioCompletion::complete``` would set the state to CALLBACK before invoking the callbacks and would set it to COMPLETE before signaling the condition.

**Comment by marcan** [comment]
Done, PTAL.

File: src/librbd/AioCompletion.h [hunk 1]
```diff
@@ -46,7 +52,7 @@ namespace librbd {
   struct AioCompletion {
     Mutex lock;
     Cond cond;
-    bool done;
+    aio_state_t done;
```
Comment (dillaman) [inline src/librbd/AioCompletion.h]: Minor: can you rename this to ```state```?

File: src/librbd/AioCompletion.cc [hunk 2]
```diff
@@ -95,12 +95,13 @@ namespace librbd {
...
     }
 
-    done = true;
+    done = STATE_CALLBACK;
     if (complete_cb) {
       lock.Unlock();
       complete_cb(rbd_comp, complete_arg);
       lock.Lock();
     }
+    done = STATE_COMPLETE;
```
Comment (dillaman) [inline src/librbd/AioCompletion.cc]: Minor: I would move this to just above ```cond.Signal()``` below

File: src/librbd/AioCompletion.h (hunk 1 above)
Comment (marcan) [inline src/librbd/AioCompletion.h]: Done.

File: src/librbd/AioCompletion.cc (hunk 2 above)
Comment (marcan) [inline src/librbd/AioCompletion.cc]: Done.

**Comment by dillaman** [comment]
lgtm

**Comment by yuriw** [comment]
test run http://pulpito.ceph.com/yuriw-2016-05-26_19:14:40-rbd-wip-yuri-testing2-distro-basic-smithi/

**Pull request merged**